pkg/frontend/querymiddleware/retry.go

```diff
@@ -71,6 +81,10 @@ func (r retry) Do(ctx context.Context, req Request) (Response, error) {
 			return resp, nil
 		}
 
+		// Any error from the Prometheus API is non-retryable.
+		if apierror.IsAPIError(err) || errors.Is(err, context.Canceled) {
```

Comment:
I see that one of the apiError types is `TypeInternal`, can you explain in this comment why that one is also non-retryable?

Reply:
Well, this is complicated...
`TypeInternal` is mapped from Prometheus `errorInternal`.
Which is used in only a few places that we're interested in, but specifically `querier.NewErrorTranslateSampleAndChunkQueryable`.

So, I guess I should exclude that.